src/opendr/perception/object_detection_2d/nanodet/algorithm/config/config_file_detail.md
# NanoDet Config File Analysis

NanoDet using [yacs](https://github.com/rbgirshick/yacs) to read YAML config file.

## Saving path

```yaml
save_dir: PATH_TO_SAVE
```

Change `save_dir` to where you want to save logs and models. If path doesn't exist, NanoDet will create it.

## Model

```yaml
model:
    arch:
        name: OneStageDetector
        backbone: xxx
        fpn: xxx
        head: xxx
```

Most detection model architecture can be divided into 3 parts: backbone, task head and connector between them (e.g., FPN, PAN).

### Backbone

```yaml
backbone:
    name: ShuffleNetV2
    model_size: 1.0x
    out_stages: [2,3,4]
    activation: LeakyReLU
    with_last_conv: False
```

NanoDet using ShuffleNetV2 as backbone. You can modify model size, output feature levels and activation function. Moreover, NanoDet provides other lightweight backbones like **GhostNet** and **MobileNetV2**. You can also add your backbone network by importing it in `nanodet/model/backbone/__init__.py`.

### FPN

```yaml
fpn:
    name: PAN
    in_channels: [116, 232, 464]
    out_channels: 96
    start_level: 0
    num_outs: 3
```

NanoDet using modified [PAN](http://arxiv.org/abs/1803.01534) (replace downsample convs with interpolation to reduce amount of computations).

`in_channels`: a list of feature map channels extracted from backbone.

`out_channels`: output feature map channel.

### Head

```yaml
head:
    name: NanoDetHead
    num_classes: 80
    input_channel: 96
    feat_channels: 96
    stacked_convs: 2
    share_cls_reg: True
    octave_base_scale: 8
    strides: [8, 16, 32]
    reg_max: 7
    norm_cfg:
      type: BN
    loss:
```

`name`: task head class name

`num_classes`: number of classes

`input_channel`: input feature map channel

`feat_channels`: channel of task head convs

`stacked_convs`: how many conv blocks use in one task head

`strides`: down sample stride of each feature map level

`reg_max`: max value of per-level l-r-t-b distance

`norm_cfg`: normalization layer setting

`loss`: adjust loss functions and weights

`assigner_cfg`: config dictionary of the assigner.

`share_cls_reg`: use same conv blocks for classification and box regression. Used in GFLHead and NanoDetHead.

`octave_base_scale`: base box scale. Used in GFLHead and NanoDetHead.

`use_depthwise`: whether to use PointWise-DepthWise or Base convolutions modules. Used in NanoDetHead and NanoDetPlusHead

`kernel_size`: size of the convolving kernel. Used in NanoDetPlusHead

`activation`: type of activation function. Used in NanoDetHead and NanoDetPlusHead

`legacy_post_process`: whether to use legacy post-processing or not.
If set to False, a faster implementation of post-processing will be used with respect to dynamic input.
Most applications will run the same with either post-processing implementations. Used in NanoDetPlusHead.

## Weight averaging

Nanodet supports weight averaging method like EMA:

```yaml
model:
  weight_averager:
    name: ExpMovingAverager
    decay: 0.9998
  arch:
    ...
```

## Data

```yaml
data:
    train:
        input_size: [320,320]
        keep_ratio: True
        cache_images: _
        multi_scale: [0.6, 1.4]
        pipeline:
    val:
        ...
```

In `data` you need to set your train and validate dataset.

`input_size`: [width, height]

`keep_ratio`: whether to maintain the original image ratio when resizing to input size.

`cache_images`: whether to cache images or not during training. "disk" option will cache images as numpy files in disk, "ram" option will cache dataset into ram.

`multi_scale`: scaling range for multi-scale training. Set to None to turn off.

`pipeline`: data preprocessing and augmentation pipeline.

## Device

```yaml
device:
    gpu_ids: [0]
    workers_per_gpu: 12
    batchsize_per_gpu: 160
    effective_batchsize: 1
```

`gpu_ids`: CUDA device id.

`workers_per_gpu`: how many dataloader processes for each gpu

`batchsize_per_gpu`: amount of images in one batch for each gpu, if -1 autobatch will determine the batchsize to be used.

`effective_batchsize`: determines the effective batch size by accumulating losses, 1 will use only batchsize_per_gpu.
## schedule

```yaml
schedule:
  resume: 0
  optimizer:
    name: SGD
    lr: 0.14
    momentum: 0.9
    weight_decay: 0.0001
  warmup:
    name: linear
    steps: 300
    ratio: 0.1
  total_epochs: 70
  lr_schedule:
    name: MultiStepLR
    milestones: [40,55,60,65]
    gamma: 0.1
  val_intervals: 10
```

Set training schedule.

`resume`: to restore # checkpoint, if 0 model start from random initialization

`load_model`: path to trained weight

`optimizer`: support all optimizer provided by pytorch.

You should adjust the `lr` with `batch_size`. Following linear scaling rule in paper *[Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour](https://research.fb.com/wp-content/uploads/2017/06/imagenet1kin1h5.pdf)*

`warmup`: warm up your network before training. Support `constant`, `exp` and `linear` three types of warm up.

`total_epochs`: total epochs to train

`lr_schedule`: please refer to [pytorch lr_scheduler documentation](https://pytorch.org/docs/stable/optim.html?highlight=lr_scheduler#torch.optim.lr_scheduler)

`val_intervals`: epoch interval of evaluating during training

## Evaluate

```yaml
evaluator:
  name: CocoDetectionEvaluator
  save_key: mAP
```

Currently only support COCO eval.

`save_key`: metric of best model. Support mAP, AP50, AP75....

****

`class_names`: used in visualization


CHANGELOG.md
# OpenDR Toolkit Change Log

## Version 3.0.0
Released on Dec, 4, 2023.
- New Features:  
  - Binary High Resolution Learner ([#402](https://github.com/opendr-eu/opendr/pull/402))
  - ROS2 node for EfficientLPS ([#404](https://github.com/opendr-eu/opendr/pull/404))
  - Fall and wave detection ROS nodes ([#423](https://github.com/opendr-eu/opendr/pull/423))
  - Continual SLAM: Adds a new SLAM tool for Continual SLAM ([#424](https://github.com/opendr-eu/opendr/pull/424))
  - Add RGB gesture recognition ([#436](https://github.com/opendr-eu/opendr/pull/436))
  - FSeq2-NMS ([#442](https://github.com/opendr-eu/opendr/pull/442))
  - Intent recognition tool ([#443](https://github.com/opendr-eu/opendr/pull/443))
  - Robotti human detection simulation demo ([#451](https://github.com/opendr-eu/opendr/pull/451))
  - Object Detection 2D Class Filtering ([#467](https://github.com/opendr-eu/opendr/pull/467))
  - RL-based Learner for Active Face Recognition ([#473](https://github.com/opendr-eu/opendr/pull/473))
  - YOLOv5s Inference Demo with Optimized Weights for Agricultural Use ([#476]([#](https://github.com/opendr-eu/opendr/pull/)476))
  - Adaptive HR Pose Estimation ([#479](https://github.com/opendr-eu/opendr/pull/479))
- Enhancements:
  - Wave detection demo based on pose estimation ([#394](https://github.com/opendr-eu/opendr/pull/394))
  - Facial expression recognition demo update ([#405](https://github.com/opendr-eu/opendr/pull/405))
  - Object detection 2d camera demos ([#408](https://github.com/opendr-eu/opendr/pull/408))
  - High Resolution Pose Estimation webcam demo ([#409](https://github.com/opendr-eu/opendr/pull/409))
  - ROS nodes FPS performance measurements ([#419](https://github.com/opendr-eu/opendr/pull/419))
  - Refactoring: pythonic joins in `test_clang_format.py`/`test_cppcheck.py` ([#455](https://github.com/opendr-eu/opendr/pull/455))
  - Test-tools improvement ([#456](https://github.com/opendr-eu/opendr/pull/456)) 
  - Adding prompt when transcribe with Whisper ([#462](https://github.com/opendr-eu/opendr/pull/462))
 
- Bug Fixes:
  - Fix package creator and sources ([#390](https://github.com/opendr-eu/opendr/pull/390))
  - Lightweight OpenPose tool fixes and improvements ([#392](https://github.com/opendr-eu/opendr/pull/392))
  - Fall Detection - alternative infer input ([#397](https://github.com/opendr-eu/opendr/pull/397))
  - Yolov5 training bugfix ([#401](https://github.com/opendr-eu/opendr/pull/401))
  - Fix the dependency conflict of geffnet installation ([#410](https://github.com/opendr-eu/opendr/pull/410))
  - Fix bug in GEM ROS2 node ([#420](https://github.com/opendr-eu/opendr/pull/420))
  - Fix link to nanodet documentation ([#421](https://github.com/opendr-eu/opendr/pull/421))
  - EfficientLPS panoptic segmentation coloring bug ([#426](https://github.com/opendr-eu/opendr/pull/426))
  - Bump flask from 1.1.2 to 2.3.2 ([#430](https://github.com/opendr-eu/opendr/pull/430))
  - Fix tests on master branch ([#438](https://github.com/opendr-eu/opendr/pull/438))
  - Added unzip installation as base ubuntu dependency and tool tests fixes ([#454](https://github.com/opendr-eu/opendr/pull/454))
  - Active face recognition demo and bug fixes on Face Recognition ([#459](https://github.com/opendr-eu/opendr/pull/459))
  - GPU installation fix ([#463](https://github.com/opendr-eu/opendr/pull/463))
  - Fix ROS1 nodes argparse issue with .launch files ([#465](https://github.com/opendr-eu/opendr/pull/465))
  - Minor fix on yolov5 webcam demo ([#466](https://github.com/opendr-eu/opendr/pull/466))
  - Apply cuDNN init fix to all Object Detectors 2D ([#469](https://github.com/opendr-eu/opendr/pull/469))
  - Updated test_suite_develop.yml based on latest test_suite.yml ([#471](https://github.com/opendr-eu/opendr/pull/471))
  - Fix fmpgmapping ([#472](https://github.com/opendr-eu/opendr/pull/472))
  - Synchronization and bugfixes ([#478](https://github.com/opendr-eu/opendr/pull/478))

## Version 2.2.0
Released on July, 3rd, 2023.

  - Dependency Updates ([#431](https://github.com/opendr-eu/opendr/pull/431)):
    - Updated PyTorch version from 1.9.0 to 1.13.1.
    - Updated Torchvision version from 0.10.0 to 0.14.1.
    - Updated Torchaudio version from 0.9.0 to 0.13.1.
    - Downgraded wheel version to 0.38.4 due to bug in recent version.

## Version 2.1.0
Released on February, 22nd, 2023.

  - New Features:
    - Added Efficient LiDAR Panoptic Segmentation ([#359](https://github.com/opendr-eu/opendr/pull/359)).
    - Added Nanodet 2D Object Detection tool ([#352](https://github.com/opendr-eu/opendr/pull/352)).
    - Added C API implementations of NanoDet 2D Object Detection tool ([#352](https://github.com/opendr-eu/opendr/pull/352)).
    - Added C API implementations of forward pass of DETR 2D Object Detection tool ([#383](https://github.com/opendr-eu/opendr/pull/383)).
    - Added C API implementations of forward pass of DeepSORT 2D Object Tracking tool ([#383](https://github.com/opendr-eu/opendr/pull/383)).
    - Added C API implementations of forward pass of Lightweight OpenPose, Pose Estimator tool ([#383](https://github.com/opendr-eu/opendr/pull/383)).
    - Added C API implementations of forward pass of X3D 2D Activity Recognition tool ([#383](https://github.com/opendr-eu/opendr/pull/383)).
    - Added C API implementations of forward pass of Progressive Spatiotemporal GCN Skeleton-based Action Recognition tool ([#383](https://github.com/opendr-eu/opendr/pull/383)).
    - Added Binary High Resolution Analysis tool ([#402](https://github.com/opendr-eu/opendr/pull/402)).
    - Added Multi-Object-Search tool ([#363](https://github.com/opendr-eu/opendr/pull/363))
  - Enhancements:
    - Added support in C API for detection target structure and vector of detections ([#352](https://github.com/opendr-eu/opendr/pull/352))
    - Added support in C API for tensor structure and vector of tensors ([#383](https://github.com/opendr-eu/opendr/pull/383))

src/perception/object_tracking_2d/fair_mot/algorithm/lib/opts.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import os
import sys


class opts(object):
    def __init__(self):
        self.parser = argparse.ArgumentParser()
        # basic experiment setting
        self.parser.add_argument("task", default="mot", help="mot")
        self.parser.add_argument("--dataset", default="jde", help="jde")
        self.parser.add_argument("--exp_id", default="default")
        self.parser.add_argument("--test", action="store_true")
        # self.parser.add_argument('--load_model', default='../models/ctdet_coco_dla_2x.pth',
        # help='path to pretrained model')
        self.parser.add_argument(
            "--load_model", default="", help="path to pretrained model"
        )
        self.parser.add_argument(
            "--resume",
            action="store_true",
            help="resume an experiment. "
            "Reloaded the optimizer parameter and "
            "set load_model to model_last.pth "
            "in the exp dir if load_model is empty.",
        )

        # system
        self.parser.add_argument(
            "--gpus", default="0, 1", help="-1 for CPU, use comma for multiple gpus"
        )
        self.parser.add_argument(
            "--num_workers",
            type=int,
            default=8,
            help="dataloader threads. 0 for single-thread.",
        )
        self.parser.add_argument(
            "--not_cuda_benchmark",
            action="store_true",
            help="disable when the input size is not fixed.",
        )
        self.parser.add_argument(
            "--seed", type=int, default=317, help="random seed"
        )  # from CornerNet

        # log
        self.parser.add_argument(
            "--print_iter",
            type=int,
            default=0,
            help="disable progress bar and print to screen.",
        )
        self.parser.add_argument(
            "--hide_data_time",
            action="store_true",
            help="not display time during training.",
        )
        self.parser.add_argument(
            "--save_all", action="store_true", help="save model to disk every 5 epochs."
        )
        self.parser.add_argument(
            "--metric", default="loss", help="main metric to save best model"
        )
        self.parser.add_argument(
            "--vis_thresh", type=float, default=0.5, help="visualization threshold."
        )

        # model
        self.parser.add_argument(
            "--arch",
            default="dla_34",
            help="model architecture. Currently tested"
            "resdcn_34 | resdcn_50 | resfpndcn_34 |"
            "dla_34 | hrnet_18",
        )
        self.parser.add_argument(
            "--head_conv",
            type=int,
            default=-1,
            help="conv layer channels for output head"
            "0 for no conv layer"
            "-1 for default setting: "
            "256 for resnets and 256 for dla.",
        )
        self.parser.add_argument(
            "--down_ratio",
            type=int,
            default=4,
            help="output stride. Currently only supports 4.",
        )

        # input
        self.parser.add_argument(
            "--input_res",
            type=int,
            default=-1,
            help="input height and width. -1 for default from "
            "dataset. Will be overriden by input_h | input_w",
        )
        self.parser.add_argument(
            "--input_h",
            type=int,
            default=-1,
            help="input height. -1 for default from dataset.",
        )
        self.parser.add_argument(
            "--input_w",
            type=int,
            default=-1,
            help="input width. -1 for default from dataset.",
        )

        # train
        self.parser.add_argument(
            "--lr", type=float, default=1e-4, help="learning rate for batch size 12."
        )
        self.parser.add_argument(
            "--lr_step", type=str, default="20", help="drop learning rate by 10."
        )
        self.parser.add_argument(
            "--num_epochs", type=int, default=30, help="total training epochs."
        )
        self.parser.add_argument(
            "--batch_size", type=int, default=12, help="batch size"
        )
        self.parser.add_argument(
            "--master_batch_size",
            type=int,
            default=-1,
            help="batch size on the master gpu.",
        )
        self.parser.add_argument(
            "--num_iters", type=int, default=-1, help="default: #samples / batch_size."
        )
        self.parser.add_argument(
            "--val_intervals",
            type=int,
            default=5,
            help="number of epochs to run validation.",
        )
        self.parser.add_argument(
            "--trainval",
            action="store_true",
            help="include validation in training and " "test on test set",
        )

        # test
        self.parser.add_argument(
            "--K", type=int, default=500, help="max number of output objects."
        )
        self.parser.add_argument(
            "--not_prefetch_test",
            action="store_true",
            help="not use parallal data pre-processing.",
        )
        self.parser.add_argument(
            "--fix_res",
            action="store_true",
            help="fix testing resolution or keep " "the original resolution",
        )
        self.parser.add_argument(
            "--keep_res",
            action="store_true",
            help="keep the original resolution" " during validation.",
        )
        # tracking
        self.parser.add_argument("--test_mot16", default=False, help="test mot16")
        self.parser.add_argument("--val_mot15", default=False, help="val mot15")
        self.parser.add_argument("--test_mot15", default=False, help="test mot15")
        self.parser.add_argument(
            "--val_mot16", default=False, help="val mot16 or mot15"
        )
        self.parser.add_argument("--test_mot17", default=False, help="test mot17")
        self.parser.add_argument("--val_mot17", default=False, help="val mot17")
        self.parser.add_argument("--val_mot20", default=False, help="val mot20")
        self.parser.add_argument("--test_mot20", default=False, help="test mot20")
        self.parser.add_argument("--val_hie", default=False, help="val hie")
        self.parser.add_argument("--test_hie", default=False, help="test hie")
        self.parser.add_argument(
            "--conf_thres",
            type=float,
            default=0.4,
            help="confidence thresh for tracking",
        )
        self.parser.add_argument(
            "--det_thres",
            type=float,
            default=0.3,
            help="confidence thresh for detection",
        )
        self.parser.add_argument(
            "--nms_thres", type=float, default=0.4, help="iou thresh for nms"
        )
        self.parser.add_argument(
            "--track_buffer", type=int, default=30, help="tracking buffer"
        )
        self.parser.add_argument(
            "--min-box-area", type=float, default=100, help="filter out tiny boxes"
        )
        self.parser.add_argument(
            "--input-video",
            type=str,
            default="../videos/MOT16-03.mp4",
            help="path to the input video",
        )
        self.parser.add_argument(
            "--output-format", type=str, default="video", help="video or text"
        )
        self.parser.add_argument(
            "--output-root",
            type=str,
            default="../demos",
            help="expected output root path",
        )

        # mot
        self.parser.add_argument(
            "--data_cfg",
            type=str,
            default="../src/lib/cfg/data.json",
            help="load data from cfg",
        )
        self.parser.add_argument(
            "--data_dir", type=str, default="/data/yfzhang/MOT/JDE"
        )

        # loss
        self.parser.add_argument(
            "--mse_loss",
            action="store_true",
            help="use mse loss or focal loss to train " "keypoint heatmaps.",
        )

        self.parser.add_argument(
            "--reg_loss", default="l1", help="regression loss: sl1 | l1 | l2"
        )
        self.parser.add_argument(
            "--hm_weight",
            type=float,
            default=1,
            help="loss weight for keypoint heatmaps.",
        )
        self.parser.add_argument(
            "--off_weight",
            type=float,
            default=1,
            help="loss weight for keypoint local offsets.",
        )
        self.parser.add_argument(
            "--wh_weight",
            type=float,
            default=0.1,
            help="loss weight for bounding box size.",
        )
        self.parser.add_argument(
            "--id_loss", default="ce", help="reid loss: ce | triplet"
        )
        self.parser.add_argument(
            "--id_weight", type=float, default=1, help="loss weight for id"
        )
        self.parser.add_argument(
            "--reid_dim", type=int, default=128, help="feature dim for reid"
        )
        self.parser.add_argument(
            "--ltrb", default=True, help="regress left, top, right, bottom of bbox"
        )

        self.parser.add_argument(
            "--norm_wh",
            action="store_true",
            help="L1(\hat(y) / y, 1) or L1(\hat(y), y)",
        )
        self.parser.add_argument(
            "--dense_wh",
            action="store_true",
            help="apply weighted regression near center or "
            "just apply regression on center point.",
        )
        self.parser.add_argument(
            "--cat_spec_wh",
            action="store_true",
            help="category specific bounding box size.",
        )
        self.parser.add_argument(
            "--not_reg_offset", action="store_true", help="not regress local offset."
        )

    def parse(self, args=""):
        if args == "":
            opt = self.parser.parse_args()
        else:
            opt = self.parser.parse_args(args)

        opt.gpus_str = opt.gpus
        opt.gpus = [int(gpu) for gpu in opt.gpus.split(",")]
        opt.gpus = [i for i in range(len(opt.gpus))] if opt.gpus[0] >= 0 else [-1]
        opt.lr_step = [int(i) for i in opt.lr_step.split(",")]

        opt.fix_res = not opt.keep_res
        print("Fix size testing." if opt.fix_res else "Keep resolution testing.")
        opt.reg_offset = not opt.not_reg_offset

        if opt.head_conv == -1:  # init default head_conv
            opt.head_conv = 256 if "dla" in opt.arch else 256
        opt.pad = 31
        opt.num_stacks = 1

        if opt.trainval:
            opt.val_intervals = 100000000

        if opt.master_batch_size == -1:
            opt.master_batch_size = opt.batch_size // len(opt.gpus)
        rest_batch_size = opt.batch_size - opt.master_batch_size
        opt.chunk_sizes = [opt.master_batch_size]
        for i in range(len(opt.gpus) - 1):
            slave_chunk_size = rest_batch_size // (len(opt.gpus) - 1)
            if i < rest_batch_size % (len(opt.gpus) - 1):
                slave_chunk_size += 1
            opt.chunk_sizes.append(slave_chunk_size)
        print("training chunk_sizes:", opt.chunk_sizes)

        opt.root_dir = os.path.join(os.path.dirname(__file__), "..", "..")
        opt.exp_dir = os.path.join(opt.root_dir, "exp", opt.task)
        opt.save_dir = os.path.join(opt.exp_dir, opt.exp_id)
        opt.debug_dir = os.path.join(opt.save_dir, "debug")
        print("The output will be saved to ", opt.save_dir)

        if opt.resume and opt.load_model == "":
            model_path = (
                opt.save_dir[:-4] if opt.save_dir.endswith("TEST") else opt.save_dir
            )
            opt.load_model = os.path.join(model_path, "model_last.pth")
        return opt

    def update_dataset_info_and_set_heads(self, opt, dataset):
        input_h, input_w = dataset.default_resolution
        opt.mean, opt.std = dataset.mean, dataset.std
        opt.num_classes = dataset.num_classes

        # input_h(w): opt.input_h overrides opt.input_res overrides dataset default
        input_h = opt.input_res if opt.input_res > 0 else input_h
        input_w = opt.input_res if opt.input_res > 0 else input_w
        opt.input_h = opt.input_h if opt.input_h > 0 else input_h
        opt.input_w = opt.input_w if opt.input_w > 0 else input_w
        opt.output_h = opt.input_h // opt.down_ratio
        opt.output_w = opt.input_w // opt.down_ratio
        opt.input_res = max(opt.input_h, opt.input_w)
        opt.output_res = max(opt.output_h, opt.output_w)

        if opt.task == "mot":
            opt.heads = {
                "hm": opt.num_classes,
                "wh": 2 if not opt.ltrb else 4,
                "id": opt.reid_dim,
            }
            if opt.reg_offset:
                opt.heads.update({"reg": 2})
            opt.nID = dataset.nID
            opt.img_size = (1088, 608)
            # opt.img_size = (864, 480)
            # opt.img_size = (576, 320)
        else:
            assert 0, "task not defined!"
        print("heads", opt.heads)
        return opt

    def init(self, args=""):
        default_dataset_info = {
            "mot": {
                "default_resolution": [608, 1088],
                "num_classes": 1,
                "mean": [0.408, 0.447, 0.470],
                "std": [0.289, 0.274, 0.278],
                "dataset": "jde",
                "nID": 14455,
            },
        }

        class Struct:
            def __init__(self, entries):
                for k, v in entries.items():
                    self.__setattr__(k, v)

        opt = self.parse(args)
        dataset = Struct(default_dataset_info[opt.task])
        opt.dataset = dataset.dataset
        opt = self.update_dataset_info_and_set_heads(opt, dataset)
        return opt
